framework/src/main/java/org/checkerframework/framework/type/QualifierHierarchy.java

```diff
@@ -103,34 +117,142 @@ default int getWidth() {
    * @return {@code true} if the qualifier is a polymorphic qualifier; otherwise, returns {@code
    *     false}.
    */
-  boolean isPolymorphicQualifier(AnnotationMirror qualifier);
+  public abstract boolean isPolymorphicQualifier(AnnotationMirror qualifier);
 
   // **********************************************************************
   // Qualifier Hierarchy Queries
   // **********************************************************************
 
   /**
    * Tests whether {@code subQualifier} is equal to or a sub-qualifier of {@code superQualifier},
-   * according to the type qualifier hierarchy.
+   * according to the type qualifier hierarchy, ignoring Java basetypes.
+   *
+   * <p>Clients should generally call {@link #isSubtypeShallow}. However, subtypes should generally
+   * override this method (if needed).
+   *
+   * @param subQualifier possible subqualifier
+   * @param superQualifier possible superqualifier
+   * @return true iff {@code subQualifier} is a subqualifier of, or equal to, {@code superQualifier}
+   */
+  protected abstract boolean isSubtypeQualifiers(
+      AnnotationMirror subQualifier, AnnotationMirror superQualifier);
+
+  /**
+   * Tests whether {@code subQualifier} is equal to or a sub-qualifier of {@code superQualifier},
+   * according to the type qualifier hierarchy, ignoring Java basetypes.
+   *
+   * <p>This method just calls {@code isSubtypeQualifiers()}. However, {@code isSubtypeQualifiers()}
+   * is not public. A client that wishes to call {@code isSubtypeQualifiers()} should instead call
+   * this method. This makes it obvious where code outside QualifierHierarchy is ignoring Java
+   * basetypes.
+   *
+   * @param subQualifier possible subqualifier
+   * @param superQualifier possible superqualifier
+   * @return true iff {@code subQualifier} is a subqualifier of, or equal to, {@code superQualifier}
+   */
+  public final boolean isSubtypeQualifiersOnly(
+      AnnotationMirror subQualifier, AnnotationMirror superQualifier) {
+    return isSubtypeQualifiers(subQualifier, superQualifier);
+  }
+
+  /**
+   * Returns the qualifier that is the lowest in the hierarchy. If the two qualifiers are not
+   * comparable, then the qualifier that is first when compared with {@link
+   * AnnotationUtils#compareAnnotationMirrors(AnnotationMirror, AnnotationMirror)}. then the
+   * qualifier that is first in A
+   *
+   * @param qual1 a qualifier
+   * @param qual2 a qualifier
+   * @return the qualifier that is the lowest in the hierarchy.
+   */
+  public final AnnotationMirror lowestQualifier(AnnotationMirror qual1, AnnotationMirror qual2) {
```

Comment:
This is only used in CFAbstractValue.  The other places that I thought it would be useful had different semantics.  So, it might make sense to move this method there.